Examples/Io/Json/src/JsonSurfacesWriter.cpp

```diff
@@ -0,0 +1,131 @@
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "ActsExamples/Io/Json/JsonSurfacesWriter.hpp"
+
+#include "Acts/Geometry/GeometryHierarchyMap.hpp"
+#include "Acts/Geometry/TrackingVolume.hpp"
+#include "Acts/Plugins/Json/GeometryHierarchyMapJsonConverter.hpp"
+#include "Acts/Plugins/Json/SurfaceJsonConverter.hpp"
+#include "Acts/Surfaces/Surface.hpp"
+#include "ActsExamples/Utilities/Paths.hpp"
+
+using namespace ActsExamples;
+
+JsonSurfacesWriter::JsonSurfacesWriter(const JsonSurfacesWriter::Config& cfg,
+                                       Acts::Logging::Level lvl)
+    : m_cfg(cfg),
+      m_world(nullptr),
+      m_logger(Acts::getDefaultLogger("JsonSurfacesWriter", lvl)) {
+  if (not m_cfg.trackingGeometry) {
+    throw std::invalid_argument("Missing tracking geometry");
+  }
+  m_world = m_cfg.trackingGeometry->highestTrackingVolume();
+  if (not m_world) {
+    throw std::invalid_argument("Could not identify the world volume");
+  }
+}
+
+std::string JsonSurfacesWriter::name() const {
+  return "JsonSurfacesWriter";
+}
+
+namespace {
+
+using SurfaceContainer =
+    Acts::GeometryHierarchyMap<std::shared_ptr<const Acts::Surface>>;
+using SurfaceConverter = Acts::GeometryHierarchyMapJsonConverter<
+    std::shared_ptr<const Acts::Surface>>;
+
+/// Write all child surfaces and descend into confined volumes.
+void collectSurfaces(std::vector<SurfaceContainer::InputElement>& cSurfaces,
+                     const Acts::TrackingVolume& volume, bool writeLayer,
+                     bool writeApproach, bool writeSensitive,
+                     bool writeBoundary) {
+  // Process all layers that are directly stored within this volume
+  if (volume.confinedLayers()) {
+    for (auto layer : volume.confinedLayers()->arrayObjects()) {
+      // We jump navigation layers
+      if (layer->layerType() == Acts::navigation) {
+        continue;
+      }
+      // Layer surface
+      if (writeLayer) {
+        auto layerSurfacePtr = layer->surfaceRepresentation().getSharedPtr();
+        cSurfaces.push_back(SurfaceContainer::InputElement{
+            layer->surfaceRepresentation().geometryId(), layerSurfacePtr});
+      }
+      // Approach surfaces
+      if (writeApproach and layer->approachDescriptor()) {
+        for (auto sf : layer->approachDescriptor()->containedSurfaces()) {
+          cSurfaces.push_back(SurfaceContainer::InputElement{
+              sf->geometryId(), sf->getSharedPtr()});
+        }
+      }
+      // Check for sensitive surfaces
+      if (layer->surfaceArray() and writeSensitive) {
+        for (auto surface : layer->surfaceArray()->surfaces()) {
+          if (surface) {
+            cSurfaces.push_back(
+                SurfaceContainer::InputElement{surface->geometryId(), surface});
+          }
+        }
+      }
+    }
+    // This is a navigation volume, write the boundaries
+    if (writeBoundary) {
+      for (auto bsurface : volume.boundarySurfaces()) {
+        // ;
```

Comment:
That's a place-holder, but I think I can add it. For writing the boundaries, I should actually resolve also the volume pointing, but then this would be a different type ... I will add "just" the plain surface information here.